cleanlab/object_detection/filter.py

```diff
@@ -217,3 +234,146 @@ def _pool_box_scores_per_image(is_issue_per_box: List[np.ndarray]) -> np.ndarray
         if np.sum(issue_per_box) > 0:
             is_issue[idx] = 1
     return is_issue
+
+
+def _process_class_list(class_list: List[np.ndarray], class_dict: Dict[int, float]):
+    """
+    Converts a list of classes using a float dictionary into a list where each class is replaced by its corresponding float value.
+    """
+    class_l2 = []
+    for i in class_list:
+        l3 = [class_dict[j] for j in i]
+        class_l2.append(l3)
+    return class_l2
+
+
+def calculate_ap_per_class(
+    labels: List[Dict[str, Any]],
+    predictions: List[np.ndarray],
+    iou_threshold: float = 0.5,
+    num_procs: int = 4,
+):
+    """
+    Computes the average precision for each class based on provided labels and predictions.
+    It uses an Intersection over Union (IoU) threshold and supports parallel processing with a specified number of processes.
+
+    """
+    num_imgs = len(predictions)
+    num_scales = 1
+    num_classes = len(predictions[0])
+    if num_imgs > 1:
+        num_procs = min(num_procs, num_imgs)
+        pool = Pool(num_procs)
+    ap_per_class_list = []
+    for class_num in range(num_classes):
+        cls_dets, cls_gts = filter_by_class(labels, predictions, class_num)
+        if num_imgs > 1:
+            tpfp = pool.starmap(
+                _get_tp_fp,
+                zip(cls_dets, cls_gts, [iou_threshold for _ in range(num_imgs)]),
+            )
+        else:
+            tpfp = [
+                _get_tp_fp(
+                    cls_dets[0],
+                    cls_gts[0],
+                    iou_threshold,
+                )
+            ]
+        tp, fp = tuple(zip(*tpfp))
+        num_gts = np.zeros(num_scales, dtype=int)
+        for j, bbox in enumerate(cls_gts):
+            num_gts[0] += bbox.shape[0]
+        cls_dets = np.vstack(cls_dets)
+        sort_inds = np.argsort(-cls_dets[:, -1])
+        tp = np.hstack(tp)[:, sort_inds]
+        fp = np.hstack(fp)[:, sort_inds]
+        tp = np.cumsum(tp, axis=1)
+        fp = np.cumsum(fp, axis=1)
+        eps = np.finfo(np.float32).eps
+        recalls = tp / np.maximum(num_gts[:, np.newaxis], eps)
+        precisions = tp / np.maximum((tp + fp), eps)
+        recalls = recalls[0, :]
+        precisions = precisions[0, :]
+        ap = calculate_average_precision(recalls, precisions)
+        ap_per_class_list.append(ap)
+    if num_imgs > 1:
+        pool.close()
+    return ap_per_class_list
+
+
+def filter_by_class(labels: List[Dict[str, Any]], predictions: List[np.ndarray], class_num: int):
+    """
+    Filters predictions and labels based on a specific class number.
+    """
+    pred_bboxes = [img_res[class_num] for img_res in predictions]
```

Comment:
what is `img_res`? Wouldn't this be `prediction`

Reply:
https://github.com/cleanlab/cleanlab/blob/3024fad44f236c2f546610be35c3cfa5d011ce31/cleanlab/object_detection/filter.py#L312